runtime/binding-http/src/main/java/io/aklivity/zilla/runtime/binding/http/internal/stream/HttpServerFactory.java

```diff
@@ -5057,21 +5099,30 @@ private int onDecodeData(
 
                     if (payloadLength > 0)
                     {
-                        payloadRemaining.set(payloadLength);
-                        exchange.doRequestData(traceId, payload, payloadRemaining);
-                        progress += payloadLength - payloadRemaining.value;
-                        deferred += payloadRemaining.value;
-                    }
-
-                    if (deferred == 0 && Http2Flags.endStream(flags))
-                    {
-                        if (exchange.requestContentLength != -1 && exchange.contentObserved != exchange.requestContentLength)
+                        boolean isValid = binding.validateContent(exchange.request, payload, 0, payloadLength);
+                        if (isValid)
                         {
-                            doEncodeRstStream(traceId, streamId, Http2ErrorCode.PROTOCOL_ERROR);
+                            payloadRemaining.set(payloadLength);
+                            exchange.doRequestData(traceId, payload, payloadRemaining);
+                            progress += payloadLength - payloadRemaining.value;
+                            deferred += payloadRemaining.value;
+                            if (deferred == 0 && Http2Flags.endStream(flags))
+                            {
+                                if (exchange.requestContentLength != -1 &&
+                                    exchange.contentObserved != exchange.requestContentLength)
+                                {
+                                    doEncodeRstStream(traceId, streamId, Http2ErrorCode.PROTOCOL_ERROR);
+                                }
+                                else
+                                {
+                                    exchange.doRequestEnd(traceId, EMPTY_OCTETS);
+                                }
+                            }
                         }
                         else
                         {
-                            exchange.doRequestEnd(traceId, EMPTY_OCTETS);
+                            doEncodeHeaders(traceId, authorization, streamId, headers400, true);
+                            exchange.doRequestAbort(traceId, EMPTY_OCTETS);
```

Comment:
This code restructuring prevents validation of empty payload and treats empty payload as a special case.
Let's keep the simpler codepath where we don't introduce the special case for empty payload.

Reply:
done